### PR TITLE
feat(viya): while build|web|deploy - changes related to viya file serving

### LINKED ISSUE #707: sasjs web optimisations for Viya
Our current approach for streaming apps involves compiling each web file (CSS, JS, HTML, PNG etc) as a streaming service, SAS 9 STP or Viya.

With Viya however we actually have the ability to load these files as _files_ rather than services.  This reduces complexity, aids debugging, and most importantly - means that we can retain a fixed URL when opening the homepage (rather than a temporary job UUID).

The following parts need to be changed:

## Deploy Servicepack
This needs to be able to deploy non ".sas" files as native content using the files API

## sasjs web
When compiling the clickme page, for Viya, we need to change the underlying urls to `_file=` rather than `_program=`.  The default behaviour of redirecting the output via the `/files/files/UUID/content` endpoint can be prevented by adding the `_debug=2` url parameter.  This will serve the content through an iframe, preserving the URL.  To fix the dimensions of the iframe, we will need to inject the following code in the HTML page:

```html
<script>
window.frameElement.style="height:100%;width:100%;";
window.frameElement.setAttribute("allowfullscreen","")
window.frameElement.setAttribute("frameborder","0")
window.frameElement.setAttribute("marginheight","0")
window.frameElement.setAttribute("marginwidth","0")
window.frameElement.setAttribute("scrolling","auto")
</script>
```

## sasjs compile (build.sas program)
We need to add a macro to the precode ([mv_createfile.sas](https://core.sasjs.io/mv__createfile_8sas.html)) and invoke it (instead of [mv_createwebservice.sas](https://core.sasjs.io/mv__createwebservice_8sas.html) eg as follows:

`%mv_createfile(path=$(appLoc)/services/web,name=somefile.js,inref=sasjs)`

## Log message

At the end of the build.sas program, rather than directing users to the `$url/SASJobExecution/?_program=/appLoc/services/clickme` service, we should send them to `$url/SASJobExecution/?_file=/appLoc/services/clickme&_debug=2`

### PR DESCRIPTION
**Pull request by saadjutt01**
## Issue

Closes #707 

## Intent

- Create `clickme.html`
- Keep original stylesheets and script files
- Appropriate references in `clickme.html` file
- All these file should be in servicePack with `type: 'file'`
- Make appropriate changes in `build.sas` using `mv_createFile macro`

## Implementation

Changes in `web/index.ts`
- aded `adjustIframeScript` for using debug=2 in url
- updated `updateFaviconHref` to copy actual file to `webv` in case of VIYA
- added `createClickMeFile` makes URL changes  and create `index.html` file instead of service
- updated `assetPathMap` to contain filePaths of files copied to `webv` in case of VIYA

Changes in  `build.ts`
- updated function `getBuildInfo` to have macro `viya/mv_createfile.sas` and it's dependency.
- updated function `getWebServiceScriptInvocation` using `mv_createfile` in case of SASViya and file is non-sas.
- added function `getFileText` similar to `getServiceText` in case of non-sas file 
- - which returns code for `bulid.sas` using `mv_createfile` macro
- - which also read file content as base64 encoding for binary files.

**ServicePack Structure:**
Sas Member
```
{
       "name": "appinit",
       "type": "service",
       "code": " . . . "
}
```

Non Sas Member
```
{
       "name": "index.html",
       "type": "file",
       "path": " . . . "
}
```

Changes in `deploy.ts`
- added function `populateCodeInServicePack`, to populate content of binary files for servicePack

## Checks

- [x] Code is formatted correctly (`npm run lint:fix`).
image
- [x] Any new functionality has been unit tested.
- [x] All unit tests are passing (`npm test`).
Mocked:
image
Server:
image
- [x] All CI checks are green.
- [ ] JSDoc comments have been added or updated.
- [x] Reviewer is assigned.

### CODE REVIEW

**Review by allanbowe — DISMISSED** [review]

**Review by VladislavParhomchik — APPROVED** [review]

**Review by allanbowe — APPROVED** [review]

**Review by allanbowe — DISMISSED** [review]

**Review by allanbowe — APPROVED** [review]

File: src/commands/web/spec/web.spec.ts [hunk 1]
```diff
@@ -54,7 +72,34 @@ describe('sasjs web', () => {
...
+      await expect(verifyFolder(webBuiltFilesSASVIYA)).resolves.toEqual(true)
+    },
+    2 * 60 * 1000
+  )
+
+  it(
+    `should create web app with minimal template for target SAS9`,
+    async () => {
+      appName = `cli-test-web-minimal-${generateTimestamp()}`
+      await createTestMinimalApp(__dirname, appName)
```
Comment (YuryShkoda) [inline src/commands/web/spec/web.spec.ts:84]: You shouldn't create new minimal app for each test suite

**Review by YuryShkoda — CHANGES_REQUESTED** [review]

File: src/commands/build/build.ts [hunk 2]
```diff
@@ -142,12 +157,34 @@ async function getCreateWebServiceScript(serverType: ServerType) {
...
+  encoded: boolean = false
+) {
   const loc = filePath === '' ? 'services' : 'tests'
 
   switch (serverType) {
     case ServerType.SasViya:
-      return `%mv_createwebservice(path=&appLoc/${loc}/&path, name=&service, code=sascode ,replace=yes)`
+      const encodedParam = encoded ? ', intype=BASE64' : ''
+      return isSASFile
+        ? `%mv_createwebservice(path=&appLoc/${loc}/&path, name=&service, code=sascode ,replace=yes)`
```
[the remark this answers is not in the document]
Comment (allanbowe) [reply src/commands/build/build.ts:186]: It *is* a variable - a SAS Macro variable (it's set on line 2 of the generated build.sas file)

File: src/commands/build/build.ts [hunk 3]
```diff
@@ -290,6 +340,72 @@ filename sascode clear;
...
+${getWebServiceScriptInvocation(serverType, undefined, false, encoded)}
+filename filecode clear;
+`
+}
+
+async function getWebFileContent(filePath: string, type: string) {
+  let lines,
+    encoded = false
+
+  const typesToEncode: string[] = ['ICO', 'PNG', 'JPG', 'JPEG', 'SVG', 'MP3']
```
Comment (allanbowe) [inline src/commands/build/build.ts:371]: isn't SVG always text?  no harm to encode I guess

File: src/commands/web/spec/web.spec.ts (hunk 1 above)
Comment (allanbowe) [reply src/commands/web/spec/web.spec.ts:84]: Is this for performance?  Yeah I guess you could test both SAS 9 and Viya in the same test

File: src/commands/web/spec/web.spec.ts [hunk 4]
```diff
@@ -54,7 +72,34 @@ describe('sasjs web', () => {
...
+          streamWeb: true,
+          streamWebFolder: 'webv',
+          webSourcePath: 'src',
+          streamServiceName: 'clickme'
+        }
+      })
+      await updateTarget(
+        { serverUrl: undefined, streamConfig: undefined },
+        'sas9'
+      )
```
Comment (allanbowe) [inline src/commands/web/spec/web.spec.ts:98]: this is also testing that streamconfig works at root level right? Nice

**Review by allanbowe — APPROVED** [review]

File: src/commands/build/build.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (YuryShkoda) [reply src/commands/build/build.ts:186]: got it

File: src/commands/build/build.ts [hunk 5]
```diff
@@ -222,21 +259,34 @@ async function getContentFor(
...
-    const transformedContent = getServiceText(
-      file,
-      fileContent,
-      serverType,
-      testPath
-    )
-
-    content += `\n${transformedContent}\n`
+    const filePath = path.join(folderPath, file)
+    const isSASFile = /.sas$/.test(file)
```
[the remark this answers is not in the document]
Comment (saadjutt01) [reply src/commands/build/build.ts]: updated

File: src/commands/build/build.ts [hunk 6]
```diff
@@ -290,6 +340,72 @@ filename sascode clear;
 `
 }
 
+async function getFileText(
+  fileName: string,
+  filePath: string,
+  serverType: ServerType
+) {
+  const fileExtension = fileName
+    .substring(fileName.lastIndexOf('.') + 1, fileName.length)
```
[the remark this answers is not in the document]
Comment (saadjutt01) [reply src/commands/build/build.ts]: updated

File: src/commands/web/spec/web.spec.ts (hunk 1 above)
Comment (saadjutt01) [reply src/commands/web/spec/web.spec.ts]: performing sas9 and sasviya builds on same app now

File: src/commands/build/build.ts (hunk 3 above)
Comment (saadjutt01) [reply src/commands/build/build.ts]: removed SVG from list

**Review by YuryShkoda — DISMISSED** [review]

**Review by allanbowe — APPROVED** [review]

**Pull request merged**